### PR TITLE
Mark final and unhashable classes in `types` as such

### PR DESCRIPTION
**Pull request by BvB93**
A number of classes in `types` are currently unhashable and/or final but not marked a such.
This PR fixes aforementioned issue.

### CODE REVIEW

File: stdlib/types.pyi [hunk 1]
```diff
@@ -350,6 +369,9 @@ DynamicClassAttribute = property
 
 def coroutine(func: Callable[..., Any]) -> CoroutineType: ...
 
+if sys.version_info >= (3, 8):
+    CellType = _Cell
+
```
Comment (BvB93) [inline stdlib/types.pyi:374]: `CellType` was made public in [python 3.8](https://docs.python.org/3/library/types.html#types.CellType).

File: stdlib/types.pyi [hunk 2]
```diff
@@ -366,6 +388,7 @@ if sys.version_info >= (3, 10):
     from builtins import _NotImplementedType
 
     NotImplementedType = _NotImplementedType  # noqa F811 from builtins
+    @final
     class Union:
```
Comment (JelleZijlstra) [inline stdlib/types.pyi:392]: This ended up getting renamed to `UnionType`: https://github.com/python/cpython/blob/3.10/Lib/types.py#L301. Could you change that while you're at it?

Comment (BvB93) [reply stdlib/types.pyi]: No problem; done as of https://github.com/python/typeshed/pull/5837/commits/141ca5eda375710090b80bd1583a82eff3569e4b.

**Review by JelleZijlstra — APPROVED** [review]

**Comment by JelleZijlstra** [comment]
I think you'll have to add UnionType to the stubtest allowlist (https://github.com/python/typeshed/blob/master/tests/stubtest_allowlists/py310.txt) for now; we're testing against a version of 3.10 that doesn't have the UnionType rename yet.

**Pull request merged**